packages/core/src/workflow/piece-buffering.js

```diff
@@ -1,59 +1,173 @@
+import * as Server from '@ucanto/server'
+
 import { decode as decodePiece } from '../data/piece.js'
+import { getMessageGroupId } from '../utils.js'
 
 /**
  * @typedef {import('@web3-storage/data-segment').PieceLink} PieceLink
+ * @typedef {import('../data/types.js').Piece<PieceLink>} Piece
  * @typedef {import('../data/types.js').Buffer<PieceLink>} Data
  * @typedef {import('../data/types.js').PiecePolicy} PiecePolicy
+ * 
+ * @typedef {object} EncodedRecord
+ * @property {string} body
+ * @property {string} id
+ * 
+ * @typedef {object} Record
+ * @property {Piece} piece
+ * @property {string} id
  */
 
 /**
  * @param {object} props
  * @param {import('@web3-storage/filecoin-api/types').Store<Data>} props.storeClient 
  * @param {import('@web3-storage/filecoin-api/types').Queue<Data>} props.queueClient
- * @param {string[]} props.pieceRecords 
- * @param {string} [props.groupId]
+ * @param {EncodedRecord[]} props.records - message encoded piece records
+ * @param {boolean} [props.disableMessageGroupId] - only supported in FIFO queues
  */
 export async function bufferPieces ({
   storeClient,
   queueClient,
-  pieceRecords,
-  groupId
+  records,
+  disableMessageGroupId,
```

Comment:
Unfortunately, the existing docker image we can use for testcontainers to mimic a real Queue is an elasticmq which using dynamo client to create a queue could not make it a FIFO, so that we can use message group ID. Therefore, we need to allow not using it